### DISCUSSION TITLE
How can I check if a file exists in the render directory?

### DISCUSSION THREAD

**Discussion by coatless** [quarto-dev/quarto-cli#8461] [Q&A]
### Description

I need to check whether a specific file exists in the Quarto render directory before proceeding with further actions. While I can override the file on each render, I would like to try to preserve it as users may modify it or substitute in their values. 

Given this conundrum, I'm wondering if it is okay to list the resource under `contributes` instead of trying to register it with `quarto.doc.add_supporting()`/`quarto.doc.add_format_resource(path)`/`quarto.doc.add_html_dependency()`, e.g. 

```yml
contributes:
  filters:
    - filter.lua
  resources:
    - config.js
```

https://quarto.org/docs/extensions/filters.html#filter-extensions

I only saw one example of using a `resource`-like key inside of the dev article on [Custom formats](https://quarto.org/docs/extensions/formats.html#format-resources).

**Comment by cscheid** [comment]
> I need to check whether a specific file exists in the Quarto render directory before proceeding with further actions.

Where is this check happening, and what are the further actions?

**Reply by coatless** [reply]
@cscheid The check can happen in either `function Meta()` (at the beginning) or `function Pandoc(doc)` (at the end) of the filter extension.

Further actions would be to either: read the configuration from the working directory or write (and register?) the configuration file.

**Reply by cscheid** [reply]
> read the configuration from the working directory

Who's writing that configuration into the working directory? Is the issue that you don't know the path to it? I think I'm missing something. Why can't you use Lua code for this?

**Reply by coatless** [reply]
I can use Lua code to check existence, write, and retrieve the configuration; but, I need a way to know where the path is to it.

Is there something like this that already exists? I'm aware of [`pandoc.system.get_working_directory()`](https://pandoc.org/lua-filters.html#module-pandoc.system); but, this does not stay at the Quarto directory level. I'm also aware of [`quarto.utils.resolve_path(path)`](https://quarto.org/docs/extensions/lua-api.html#utility-functions); however, this only reads internal resources from `_extensions/` folder. 

I've spun up a [sample repository](https://github.com/coatless-quarto/pathing) that uses a website-based project with a filter that outputs the above command when the file is rendered to the console. However, I'm getting the working directory of the current file Pandoc is touching.

In other words, consider the following website-based project layout:

```sh
.
|- _extensions
|- _quarto.yml
|- posts/
|   |- making-a-map/
|      |_ index.qmd
|   |- cleaning-data/
|      |_ index.qmd
|- about.qmd
|_ index.qmd
```

Then, the goal would be:

```sh
quarto.utils.project_working_directory()
```

If run from `making-a-map/index.qmd` would report either a relative path to `.` or absolute:

```
# Relative into `.`
../../

# Absolute
C:/Users/toad/quarto/pathing
```

The other part that may be useful for a fallback if `_quarto.yml` isn't found would be the location of the current Qmd being rendered (e.g. for HTML/RevealJS):

```sh
quarto.utils.file_location()
# C:/Users/toad/quarto/pathing/_posts/making-a-map
```

The latter command is a remap of [`pandoc.system.get_working_directory()`](https://pandoc.org/lua-filters.html#module-pandoc.system).

**Reply by cscheid** [reply]
Instead of writing to the output directory, would it work for you to write to the same directory as the .qmd file and then indicate the file as a resource to be included?

**Reply by coatless** [reply]
When you say: 

> the same directory as the .qmd file

Does this mean I would get the path to `posts/making-a-map/` if Quarto is rendering `posts/making-a-map/index.qmd`? This information is currently being obtained from [`pandoc.system.get_working_directory()`](https://pandoc.org/lua-filters.html#module-pandoc.system).

Would it be possible to get the main project's path to where `_quarto.yml` is?

In short, the logic here is:

1. Check for existence of `config.js`/`config.toml`/`config.json`
2. If found then read in portions of the config:
    - check if config matches latest version 
        - otherwise pull from template and move existing config to backup.
3. If not found then: 
    - copy from template into working directory via either `quarto.doc.add_supporting()`/`quarto.doc.add_format_resource(path)`
4. Retrieve values from config

> then indicate the file as a resource to be included?

Yes!

**Reply by coatless** [reply]
Side question: 

For a single Quarto render session, is there a way to "flag" or store between applications of the Pandoc Lua filter that this process has already been completed?

**Reply by coatless** [reply]
@cscheid Awesome! I think this was just added in https://github.com/quarto-dev/quarto-cli/pull/8525

I think from Lua, I can access this information with:

```sh
os.getenv("QUARTO_PROJECT_ROOT")
```

Any thoughts on the side question of being able to pass a state between applications of a filter?

**Reply by cscheid** [reply]
> For a single Quarto render session, is there a way to "flag" or store between applications of the Pandoc Lua filter that this process has already been completed?

I'm pretty sure the answer is no, because we have to run the filters every time for every document (they walk the AST and change the document, we can't assume "it's already been completed": it needs to happen every time.)

But I don't know what's "this" in "this process" in your original sentence, so I might be missing something.

**Reply by coatless** [reply]
@cscheid I'm trying to reduce needing to re-check if a configuration is present and potentially re-use a loaded config instead of re-reading from disk. 

The goal would be to only execute the following process once per `quarto render` on a project, not per-Qmd document.

1. Check for existence of `config.js`/`config.toml`/`config.json`
2. If found then read in portions of the config:
    - check if config matches latest version 
        - otherwise pull from template and move existing config to backup.
3. If not found then: 
    - copy from template into working directory via either `quarto.doc.add_supporting()`/`quarto.doc.add_format_resource(path)`
4. Retrieve values from config

**Reply by cscheid** [reply]
Ok, so Lua _filters_ are only ever called on a per-document basis. So if your document has `filters: ...` in it, that filter is necessarily (and always) called at every document render, because the goal of filters is to change _documents_.

If you have pre-render or post-render _scripts_ that are implemented in Lua, then those are not _filters_, and those get executed once at the beginning or end of a project render.

However, your comment is talking about `quarto.doc.add_supporting()`, and that call doesn't make sense from the perspective of a project, because (for example) those directories can change depending on the format being rendered, and the format is a document option, not a project option.

So I think that what you have to do is a combination of things:

- in the project pre-render script, you check for the presence of the configuration and ensure that it exists in a path that's accessible by every document render. Call that path your "cache"
- in a document filter, you always use `quarto.doc.add_supporting()` getting the values from your cache, and so on.

Does that make sense?
